ritasamir/assignment2/Q_02.java

```diff
@@ -0,0 +1,29 @@
+package assignment_2;
+
+public class Q_02 {
+
+	/**
+	 * @param head
+	 *            the root of the binary tree
+	 * @param key1
+	 * @param key2
+	 * @return node which represent the common ancestor for both key1 and key2
+	 */
+	public Node getCommonAncestor(Node head, int key1, int key2) {
+		Node node = Q_01.searchTree(head, key1);
+		// checks if the first key dosen't exist then return null
+		if (node == null) {
+			return null;
+		}
+		// for every ancestor of the key1 from bottom to up,if key2 exist on
+		// the subtree of this ancestor then it is the least common ancestor
+		while (node.getParent() != null) {

```

Comment:
Great, this is definitely much closer to the optimal solution. :) The only remaining issue is that when calling searchTree for all ancestors, we re-check the already visited sub-trees resulting in an O(n^2) solution. An easy way to fix it is to check if we've just come from that sub-tree and if so, there's no need to re-check it:

```
public Node getCommonAncestor(Node head, int key1, int key2) {
Node node = Q_01.searchTree(head, key1);
	if (node == null) {
		return null;
	}	

	Node visitedChild;
	while (node.getParent() != null) {
                visitedChild = node;
		node = node.getParent();
                // Only check a sub-tree that hasn't been checked yet.
                // Java is lazy, so if the first statement is not true, the second one won't be executed
		if (node.getLeft() != visitedChild && Q_01.searchTree(node.getLeft(), key2) != null) {
                        // found key in left sub-tree
			return node;
		} else if (node.getRight() != visitedChild && Q_01.searchTree(node.getRight(), key2) != null) {
                       // found key in right sub-tree
                       return node;
                }
	}
	return null;
}
```
I'm writing this from the top of my head, so please check if your tests still pass

Reply:
yes but we need to check if the second key is the parent of this node ,we check the left or the right sub trees but not the node itself. i added an if statement to check that, please review it